src/Propulsion.SqlStreamStore/StripedIngester.fs

```diff
@@ -0,0 +1,147 @@
+        match dict.TryGetValue key with
+        | true, value ->
+            dict.Remove key |> ignore
+            Some value
+        | false, _ -> None
+
+open StripedIngesterImpl
+
+/// Holds batches away from Core processing to limit in-flight processing
+type StripedIngester
+    (   log : ILogger, inner : Propulsion.Ingestion.Ingester<seq<StreamEvent<byte[]>>, Propulsion.Submission.SubmissionBatch<StreamEvent<byte[]>>>,
+        maxInFlightBatches, initialSeriesIndex, statsInterval : TimeSpan, ?pumpInterval) =
+    let cts = new CancellationTokenSource()
+    let pumpInterval = defaultArg pumpInterval (TimeSpan.FromMilliseconds 5.)
+    let work = ConcurrentQueue<InternalMessage>() // Queue as need ordering semantically
+    let maxInFlightBatches = Sem maxInFlightBatches
+    let stats = Stats(log, statsInterval)
+    let pending = Queue<_>()
+    let readingAhead, ready = Dictionary<int, ResizeArray<_>>(), Dictionary<int, ResizeArray<_>>()
+    let mutable activeSeries = initialSeriesIndex
+
+    let reserveAsInFlightBatch () = maxInFlightBatches.Await(cts.Token)
+    let releaseInFlightBatchAllocation () = maxInFlightBatches.Release()
+    
+    let handle = function
+        | Batch (seriesId, epoch, checkpoint, items) ->
+            let isForActiveStripe = activeSeries = seriesId
+            let batchInfo =
+                let items = Array.ofSeq items
+                let onCompleted =
+                    if isForActiveStripe then
+                        // If this read represents a batch that we will immediately submit for processing, we will defer the releasing of the batch in out buffer
+                        // limit only when the batch's processing has concluded
+                        releaseInFlightBatchAllocation
+                    else
+                        // if the batch pertains to a stripe other than the active one, we don't count that as a 'buffered item'
+                        // (there will be indirect backpressure by virtue of the fact that the processing will not mark batches on 'active' series completed until
+                        //   any ones we hold and forward through `readingAhead` are processed)
+                        // - yield a null function as the onCompleted callback to be triggered when the batch's processing has concluded
+                        id
+                epoch, checkpoint, items, onCompleted
+
+            if isForActiveStripe then
+                pending.Enqueue batchInfo
+            else
+                match readingAhead.TryGetValue seriesId with
+                | false, _ -> readingAhead.[seriesId] <- ResizeArray[|batchInfo|]
+                | true,current -> current.Add(batchInfo)
+                // As we'll be submitting `id` as the onCompleted callback, we now immediately release the allocation that gets `Await`ed in `Submit()`
+                releaseInFlightBatchAllocation()
+
+         | CloseSeries seriesIndex ->
```

Comment:
should be possible to get through both of these - it just goes straight to tail mode